Support using const pointers in asm `const` operand

nbdd0121 opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/138618)*
<!-- homu-ignore:end -->

Implements [RFC#3848](https://github.com/rust-lang/rfcs/pull/3848) with tracking issue rust-lang/rust#128464

This adds support of const pointers for asm `const` in addition to plain integers.

The inline `asm!` support is implemented using `i` constraint, and the `global_asm!` and `naked_asm!` support is implemented by inserting `symbol + offset` and make `symbol` compiler-used. For unnamed consts, it will create additional internal & hidden symbols so that they can be referenced by global_asm.

~~The feature is also implemented for GCC backend but it's untested.~~ Tested now.

nbdd0121 committed:
Implement asm_const_ptr feature

The backend now fully supports codegen of const pointers, remove the
block inside typeck behind a new feature gate. Tests are also added.

nbdd0121 committed:
Collect constants within `global_asm!` in collector

This is currently a no-op, but will be useful when const in `global_asm!`
can be pointers.

fmease commented:
r? codegen

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13906695731/job/38911409406) [(plain)](https://github.com/rust-lang/rust/commit/ffba8e8d1a8f2c88bec3d78b1a07be9484c38ede/checks/38911409406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] gccjit_sys test:false 0.024
    Checking gccjit v2.4.0
[RUSTC-TIMING] gccjit test:false 0.170
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0004]: non-exhaustive patterns: `GlobalAsmOperandRef::ConstPointer { .. }` not covered
   --> compiler/rustc_codegen_gcc/src/asm.rs:853:27
    |
853 |                     match operands[operand_idx] {
    |                           ^^^^^^^^^^^^^^^^^^^^^ pattern `GlobalAsmOperandRef::ConstPointer { .. }` not covered
    |
note: `GlobalAsmOperandRef<'_>` defined here
   --> /checkout/compiler/rustc_codegen_ssa/src/traits/asm.rs:47:1
    |
47  | pub enum GlobalAsmOperandRef<'tcx> {
```

</details>

compiler-errors commented:
r? compiler-errors

compiler-errors commented on compiler/rustc_codegen_ssa/src/traits/asm.rs:
```diff
@@ -25,9 +26,12 @@ pub enum InlineAsmOperandRef<'tcx, B: BackendTypes + ?Sized> {
         out_place: Option<PlaceRef<'tcx, B::Value>>,
     },
-    Const {
+    Interpolate {
```
Please add more doc comment explaining the significance of the difference between `Interpolate` and `Const` for backends. It begs the question for why we ever turn const operands into strings, for example.

nbdd0121 replied:
Added some explaination. Technically we should be able to use `Const` for integers, although I am a bit conservative here and don't want to change how existing things work.

This might also be useful to have in the future if we decided to add `interpolate "some CTFE string to be interpolated"` new type of operand.

compiler-errors commented on compiler/rustc_codegen_ssa/src/traits/asm.rs:
```diff
@@ -41,9 +45,20 @@ pub enum InlineAsmOperandRef<'tcx, B: BackendTypes + ?Sized> {
+        string: String,
+    },
+    ConstPointer {
+        value: mir::interpret::Pointer,
+        /// Instance that instantiates this const operand.
+        instance: Instance<'tcx>,
```
Why do we need both a value and an instance?

bors commented:
:umbrella: The latest upstream changes (presumably #140415) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14991883136/job/42117005733) [(plain)](https://github.com/rust-lang/rust/commit/c7a2301e859a394eeed9a0a1a5e105a9484ed9cc/checks/42117005733/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] cranelift_object test:false 0.628
[RUSTC-TIMING] cranelift_frontend test:false 1.013
[RUSTC-TIMING] cranelift_codegen test:false 29.962
   Compiling rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0050]: method `codegen_global_asm` has 5 parameters but the declaration in trait `rustc_codegen_ssa::traits::AsmCodegenMethods::codegen_global_asm` has 6
  --> compiler/rustc_codegen_cranelift/src/global_asm.rs:27:9
   |
27 | /         &mut self,
28 | |         template: &[InlineAsmTemplatePiece],
29 | |         operands: &[GlobalAsmOperandRef<'tcx>],
30 | |         options: InlineAsmOptions,
31 | |         _line_spans: &[Span],
   | |____________________________^ expected 6 parameters, found 5
   |
   = note: `codegen_global_asm` from trait: `fn(&mut Self, &[InlineAsmTemplatePiece], &[GlobalAsmOperandRef<'tcx>], InlineAsmOptions, &[rustc_span::Span], rustc_middle::ty::Instance<'tcx>)`

error[E0599]: no variant named `Const` found for enum `GlobalAsmOperandRef<'_>`
   --> compiler/rustc_codegen_cranelift/src/global_asm.rs:107:42
    |
107 |                     GlobalAsmOperandRef::Const { ref string } => {
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #141668) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nbdd0121 commented on tests/ui/asm/const-refs-to-static.rs:
```diff
@@ -1,19 +1,20 @@
 
-global_asm!("{}", const addr_of!(FOO));
-//~^ ERROR invalid type for `const` operand
+global_asm!("/* {} */", const addr_of!(FOO));
```
An address isn't a valid instruction

oli-obk commented on compiler/rustc_codegen_llvm/src/asm.rs:
```diff
@@ -381,13 +389,119 @@ impl<'tcx> AsmCodegenMethods<'tcx> for CodegenCx<'_, 'tcx> {
         let intel_syntax = matches!(asm_arch, InlineAsmArch::X86 | InlineAsmArch::X86_64)
             && !options.contains(InlineAsmOptions::ATT_SYNTAX);
 
+        // Convert all operands to string interpolations
```
I think it would be good to take this opportunity and move this into cg_ssa and and add some extra trait methods for the parts that are different. Similar to what I did in https://github.com/rust-lang/rust/pull/142960

nbdd0121 replied:
GCC 15+ has ability for file-level assembly to reference symbols and constants, so I think codegen_llvm and codegen_gcc may eventually diverge and use different approaches.

nbdd0121 commented:
Oops, committed submodule changes by mistake

nbdd0121 commented on tests/assembly/asm/x86-types.rs:
```diff
@@ -92,6 +92,18 @@ pub unsafe fn sym_fn() {
+// and on x86 PIC symbol can't be constant.
+// x86_64-LABEL: const_ptr:
+// x86_64: #APP
+// x86_64: mov al, byte ptr [{{.*}}anon{{.*}}]
```
`mov al, byte ptr [.Lanon.2e8c0013b01edcd3779e8174c2338a00.0]`

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16178183518/job/45668385274) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45668385274) [(plain)](https://github.com/rust-lang/rust/commit/318fd4300a8ebb29ce47ddcbb0dfc71ece183afb/checks/45668385274/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0425]: cannot find function `const_alloc_to_gcc` in module `crate::consts`
   --> compiler/rustc_codegen_gcc/src/asm.rs:923:55
    |
923 | ...                   let init = crate::consts::const_alloc_to_gcc(self, alloc);
    |                                                 ^^^^^^^^^^^^^^^^^^ not found in `crate::consts`

For more information about this error, try `rustc --explain E0425`.
[RUSTC-TIMING] rustc_codegen_gcc test:false 1.213
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
```

</details>

compiler-errors commented:
(remember to mark this as ready when it's ready 😃)

nbdd0121 commented:
I think this is ready from my side.

One thing that I am not completely certain about is on the aspect of symbol names. To be able to allow global asm to reference a promoted (unnamed) statics, I need give them names and the most sensible way to me is to use the symbol name of containing body as prefix. For naked_asm this is obvious, but for `global_asm!` this require me to give them unique names. I currently encode them like shims, I wonder that's a good-enough solution?

BTW, this also makes me wonder if we should just desugar `global_asm` as just naked functions with generated names. Given `global_asm!` have (fake) bodys now anyway, it feels like desugaring `global_asm` to `naked_asm` makes more sense then the other way around...

@rustbot ready

compiler-errors commented on compiler/rustc_symbol_mangling/src/v0.rs:
```diff
@@ -867,6 +867,16 @@ impl<'tcx> Printer<'tcx> for SymbolMangler<'tcx> {
             DefPathData::ForeignMod => return print_prefix(self),
 
+            // Global asm are handled similar to shims.
+            DefPathData::GlobalAsm => {
```
Why not just give this a prefix? Then we'll encode it with that prefix + disambiguator + empty name.

bors commented:
:pushpin: Commit 81a4b3efde1d2f4b2483d5747596f54fca214487 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 81a4b3efde1d2f4b2483d5747596f54fca214487 -->
<!-- homu: {"type":"Approved","sha":"81a4b3efde1d2f4b2483d5747596f54fca214487","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
@bors r-

486ffda9dcd0d4ef0a09d81e6ce5f241e77526a1 changed the signature for `const_alloc_to_llvm`, see the failure at https://github.com/rust-lang/rust/actions/runs/16206384671/job/45757598329?pr=143753

bors commented:
:pushpin: Commit 8da54fc3ef24b1d461524720d1ee502d27d44187 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 8da54fc3ef24b1d461524720d1ee502d27d44187 -->
<!-- homu: {"type":"Approved","sha":"8da54fc3ef24b1d461524720d1ee502d27d44187","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
@bors rollup-

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16226252611/job/45821961451)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/16226252611/job/45821961451","builder_name":"checks-actions"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #144249) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Darksonn commented:
> Looks like the i constraint doesn't work for GOT-rebased relocation

I've been playing around with this a bit, and I think the `i` operand works as long as using the symbol name would work. For example, this breaks:
```c
#include <stdio.h>

static const int FORTY_TWO = 42;

int main(void) {
    const int *a;

    __asm__ (
        "movabs %1, %0"
        : "=r" (a)
	: "i" (&FORTY_TWO)
    );

    printf("%p\n", (void *)a);

    return 0;
}
```
```text
/usr/bin/ld: /tmp/ccptrDbi.o: warning: relocation in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
```
But when we make it rip-relative, it works:
```c
#include <stdio.h>

static const int FORTY_TWO = 42;

int main(void) {
    const int *a;

    __asm__ (
        "movabs %1 - ., %0"
        : "=r" (a)
	: "i" (&FORTY_TWO)
    );

    printf("%p\n", (void *)a);

    return 0;
}
```
```text
0xec3
```
This is exactly the same scenario as when writing `FORTY_TWO` would work as far as I can tell. Is there something I'm missing?

nbdd0121 commented:
It works in x64 but not in x86.

Darksonn commented:
Ah, that's what I get for testing it with gcc.

nbdd0121 committed:
Add FnDef/Closure -> FnPtr coercion for inline asm const operand

nbdd0121 committed:
Give global_asm symbol names

Currently global_asm already have symbol names when using v0 scheme, this
makes them obtain symbols with legacy scheme too.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20272972807/job/58213822595) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58213822595) [(plain)](https://github.com/rust-lang/rust/commit/5b342bc95801e25f716df476ac2e9a66ec83e58d/checks/58213822595/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     ╭▸ compiler/rustc_hir_typeck/src/expr.rs:3701:80
     │
3701 │                     // This is mostly similar to type-checking of inline const expresssions `const { ... }`, however
     ╰╴                                                                               ━━━━━━━━━━━━
rerun tidy with `--extra-checks=spellcheck --bless` to fix typos
tidy [extra_checks]: checks with external tool 'typos' failed
tidy [extra_checks]: FAIL
tidy: The following check failed: extra_checks
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/yarn --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1612:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1358:29

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:44
  local time: Tue Dec 16 15:24:58 UTC 2025
  network time: Tue, 16 Dec 2025 15:24:58 GMT
##[error]Process completed with exit code 1.
```

</details>

nbdd0121 commented on compiler/rustc_codegen_gcc/src/asm.rs:
```diff
@@ -869,6 +890,100 @@ impl<'gcc, 'tcx> AsmCodegenMethods<'tcx> for CodegenCx<'gcc, 'tcx> {
+                        };
+
+                        let offset = offset.bytes();
+                        if offset != 0 { format!("{symbol}+{offset}") } else { symbol }
```
Hmm, `offset` is unsigned, so I elected to just emit `+{offset}`. If the offset is negative, CTFE will just produce 2's complement values as unsigned (which is okay for the wrapping semantics, and note that you have to use wrapping semantics to get pointer out of bound in the first place).

Assemblers seem to be happy with `+0xFFFFFFFFFFFFFFFF` instead of `-1`; do you think it's worth changing the generated code so we always treat offset as signed?

Darksonn replied:
IMO `-1` seems a lot safer than `+0xFFFFFFFFFFFFFFFF` since we don't have to worry about the bitwidth of the addition.

nbdd0121 committed:
Delay stringification of const to backend

nbdd0121 committed:
Unify handling of asm const and sym using CTFE

This gives the asm-const code the basic ability to deal wiht pointer and
provenances, which lays the ground work for asm_const_ptr.

Note that `SymStatic` is not fully removed, a specialized is kept and
renamed as `SymThreadLocalStatic`, for `#[thread_local]` statics where CTFE
does not support naming. The `#[thread_local]` is unstable feature and it's
not clear if we want to support this in `sym`, but removal of it should be
a separate PR.

nbdd0121 committed:
Handle pointers with offset for asm const

nbdd0121 committed:
Support codegen of asm const pointers without provenance

CTFE pointers created via type ID, `without_provenance` or pointers to const
ZSTs can now be codegenned with all 3 backends. These pointers are generated
in the same way as integers.

nbdd0121 committed:
Unify handling of `GlobalAlloc` inside backend

With the previous commit, now we can see there are some code duplication
for the handling of `GlobalAlloc` inside backends. Do some clean up to
unify them.

nbdd0121 committed:
Generate unique symbol names if const pointers refer to promoted static

nbdd0121 commented:
Quite a few changes for this new version:
* `i` operand isn't used anymore, due to the i686 limitation discussed above. The code is updated to always generate `{symbol}{offset:+}` instead (when symbol is still either interpolated directly or inserted by backend, depending on backend's capability of doing so).
* `sym fn` and const pointers are now unified, as this new approach produces identical assembly for `const f` and `sym f`.
* `sym static` is not unified. `sym STATIC` and `const &STATIC` are mostly the same, but `sym` works for `#[thread_local]` statics, while reference to `#[thread_local]` statics cannot be taken in CTFE.
* Commits are broken down into smaller (yet individually still meaningful pieces for help commit-by-commit review).

nbdd0121 commented:
r? @rust-lang/compiler

nbdd0121 commented:
Oops, sorry!

bors commented:
:umbrella: The latest upstream changes (presumably #150559) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented:
@nbdd0121 I've self-assigned this because I'm going to delta-review it this weekend, so it would be worth fixing the merge conflicts whenever convenient.

Amanieu commented on compiler/rustc_hir_typeck/src/expr.rs:
```diff
@@ -3698,7 +3698,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                    self.check_expr_const_block(anon_const, Expectation::NoExpectation);
+                    // This is mostly similar to type-checking of inline const expressions `const { ... }`, however
+                    // asm const has special coercion rules (per RFC 3848) where function items and closures are coerced to
+                    // function pointers (while pointers and integer remain as-is).
```
I would like this to be match the coercions done for in operands in check_expr_asm_operand. This means adding a coercion from references to pointers and removing closure coercion.

saethlin commented on compiler/rustc_codegen_ssa/src/mir/block.rs:
```diff
@@ -1394,14 +1397,24 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     InlineAsmOperandRef::InOut { reg, late, in_value, out_place }
                 }
                 mir::InlineAsmOperand::Const { ref value } => {
+                    let Const::Unevaluated(c, _) = &value.const_ else {
+                        bug!("need unevaluated const to derive symbol name")
+                    };
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/f82485388963286763f8c8a9261ac227627f6a18..f6a193de8f58292c28b0a702332cf56a4fff8cee)*

This seems like it might be a fiddly requirement to uphold (for example if someone tries to be clever and walk MIR and evaluate ConstOperands instead of Operands like GVN does, they'd hit this). Can you add a bit more explanation in the code for why we need the const to be unevaluated to compute a symbol name? It's not even clear to me that is a hard requirement. Conceptually, at least.

bjorn3 replied:
For regular references to consts we don't derive the symbol name from the `Instance` at all. Rather we produce an anonymous allocation, which causes LLVM to derive a name from a fixed prefix + counter. This is not globally unique, but that is fine for cg_llvm as the anonymous allocation is marked as private.

saethlin replied:
So you're saying this can/should be changed in the future?

nbdd0121 replied:
We need to generate an unique name for the allocation that the const points to, and for that name I just use the unevaluated const for now. We could perhaps do something different, e.g. create a new `DefId` for anon allocations? But it feels a huge change.

FWIW in an earlier version I just use an ad-hoc naming scheme, where the generated name is `<INSTANCE>.<INDEX NUMBER OF CONST>`. This works for `global_asm!`. However, for inline asm, this is problematic as `<INSTANCE>` is now the outer body of `asm!()` and not the `global_asm!` or `naked_asm!` itself.

bjorn3 replied:
LLVM just uses a counter for anonymous constants. You might even be able to directly create an anonymous constant and then get the name of it when necessary for global_asm.

nbdd0121 replied:
The names need to be unique if they point to different constants. Even if symbols are marked as `llvm.compiler.used`, they can still be renamed in LTO process in case there is a name conflict. Not to mention that using anon const works for LLVM but not for other CG backends.

nbdd0121 replied:
I've added a comment to explain this. Long term I think we probably still want to have a generic mechanism to ensure we can assign symbol names to anonymous allocations.

saethlin commented:
I just have one question, then r=me.

This was not an easy delta review. More fool me :joy:

bjorn3 commented:
With this PR ThinLTO would no longer be able to rename functions when making them public as the inline asm now directly embeds the symbol names, right?

nbdd0121 commented:
> With this PR ThinLTO would no longer be able to rename functions when making them public as the inline asm now directly embeds the symbol names, right?

Symbol names are handled the same way as today's `sym` operand. I.e. embedded directly inside `global_asm!` (with compiler.used added), while for inline `asm!()` it uses `s` constraint, which should still allow renaming.

bjorn3 commented on compiler/rustc_codegen_gcc/src/common.rs:
```diff
@@ -225,6 +226,76 @@ impl<'gcc, 'tcx> ConstCodegenMethods for CodegenCx<'gcc, 'tcx> {
+            let alloc = alloc.inner();
+            let typ = self.val_ty(init).get_aligned(alloc.align.bytes());
+
+            let global = self.declare_global_with_linkage(sym.name, typ, GlobalKind::Exported);
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/f82485388963286763f8c8a9261ac227627f6a18..f6a193de8f58292c28b0a702332cf56a4fff8cee)*

Exporting is wrong. That would lead to symbol conflicts if two asm blocks use the same const.

nbdd0121 replied:
Hmm, I guess changing the `Exported` to something else is not going to be sufficient and I need to also do some local dedup in case a `asm!()` is monomorphized twice due to MIR inlining? This is not an issue as the name hint is only used for `global_asm`/`naked_asm` but GCC needs injected symbol names for inline asm too.

I guess I really need to build a working version of cg_gcc and test locally.

Darksonn replied:
Could we have a test that catches this? My understanding is that Rust does run its test suite on gcc too.

nbdd0121 replied:
Yeah, I'll do that. I've already reproduced the issue locally. It looks like it has been much easier to build Rust with cg_gcc now, which isn't the case when I started working on this in 2024.

nbdd0121 replied:
Test added

bjorn3 commented on compiler/rustc_codegen_gcc/src/common.rs:
```diff
@@ -225,6 +226,76 @@ impl<'gcc, 'tcx> ConstCodegenMethods for CodegenCx<'gcc, 'tcx> {
+        let value = match alloc.inner().mutability {
+            Mutability::Mut => {
+                self.static_addr_of_mut(const_alloc_to_gcc(self, alloc), alloc.inner().align, None)
+            }
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/f82485388963286763f8c8a9261ac227627f6a18..f6a193de8f58292c28b0a702332cf56a4fff8cee)*

This feels wrong to me. consts are supposed to semantically be duplicated at every use site. Deduplication is only semantically identical in the absence of mutability.

nbdd0121 replied:
This is just moved from existing `scalar_to_backend`. Const with interior mutability cannot be promoted, so for an allocation to be mutable it has to be a `static` which is unique.

bjorn3 commented on compiler/rustc_codegen_ssa/src/traits/consts.rs:
```diff
@@ -37,6 +38,19 @@ pub trait ConstCodegenMethods: BackendTypes {
+        &self,
+        global_alloc: GlobalAlloc<'tcx>,
+        name_hint: Option<Instance<'tcx>>,
+    ) -> Result<(Self::Value, Option<Instance<'tcx>>), u64>;
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/f82485388963286763f8c8a9261ac227627f6a18..f6a193de8f58292c28b0a702332cf56a4fff8cee)*

This method is not used inside cg_ssa, so it shouldn't be defined there.

nbdd0121 replied:
It's correlated to `scalar_to_backend` and I feel this is a good place to put the doc-comments to be shared. I could move it out. Do you want me to duplicate the documentation in cg_gcc and cg_llvm?

bjorn3 commented on compiler/rustc_codegen_llvm/src/common.rs:
```diff
@@ -264,6 +264,91 @@ impl<'ll, 'tcx> ConstCodegenMethods for CodegenCx<'ll, 'tcx> {
+            GlobalAlloc::TypeId { .. } => {
+                // Drop the provenance, the offset contains the bytes of the hash, so
+                // just return 0 as base address.
+                return Err(0);
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/f82485388963286763f8c8a9261ac227627f6a18..f6a193de8f58292c28b0a702332cf56a4fff8cee)*

Could this return `Ok((LLVMConstIntToPtr(self.const_usize(0), todo!("ty")), None))`? That would allow getting rid of the `Result` returned by `alloc_to_backend`. Or does that not work because of the inbounds flag on the gep?

nbdd0121 replied:
It'll indeed be UB if GEP inbounds is used. Also, for asm_const_ptr codegen, it needs to map this to just an integer rather than a symbol + offset.

nbdd0121 committed:
Add test if asm const are MIR inlined and monomorphized twice

saethlin commented:
r? bjorn3

bjorn3 commented on compiler/rustc_codegen_cranelift/src/global_asm.rs:
```diff
@@ -137,7 +166,6 @@ fn codegen_global_asm_inner<'tcx>(
                                 "asm! and global_asm! sym operands are not yet supported",
                             );
                         }
-
```
Please restore this newline.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cf1817bc6ecd2d14ca492247c804bad31948dd56..7c877c9f0a8386b02ff1b71d2a3409533ad6e070)*

bjorn3 commented on compiler/rustc_codegen_cranelift/src/global_asm.rs:
```diff
@@ -108,27 +110,54 @@ fn codegen_global_asm_inner<'tcx>(
+                                        // current codegen unit
+                                        tcx.symbol_name(instance)
+                                    }
+                                    GlobalAlloc::Static(def_id) => {
```
What is the difference between `GlobalAsmOperandRef::Const` with a static and `GlobalAsmOperandRef::SymStatic`?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cf1817bc6ecd2d14ca492247c804bad31948dd56..7c877c9f0a8386b02ff1b71d2a3409533ad6e070)*

nbdd0121 replied:
See https://github.com/rust-lang/rust/pull/138618#issuecomment-3676004350

`#[thread_local]` works with `sym STATIC` today, but it cannot be referenced in CTFE. I would actually be in favour of removing that feature.

bjorn3 replied:
I would tend to agree with removing `#[thread_local]` support for `sym STATIC`. On some platforms (*cough* Windows *cough*) `#[thread_local]` accesses go through a function shim to ensure cross-dylib accesses work, so `sym STATIC` would probably already be broken for those platforms.

nbdd0121 replied:
Rename `GlobalAsmOperandRef::SymStatic` to `SymThreadLocalStatic`

bjorn3 commented on compiler/rustc_codegen_cranelift/src/global_asm.rs:
```diff
@@ -108,27 +110,54 @@ fn codegen_global_asm_inner<'tcx>(
+                                                span,
+                                                "asm! and global_asm! sym operands are not yet supported",
+                                            );
+                                        }
```
This error should apply to all constants that reference a symbol.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cf1817bc6ecd2d14ca492247c804bad31948dd56..7c877c9f0a8386b02ff1b71d2a3409533ad6e070)*

nbdd0121 commented:
I think the approach I use in rustc_codegen_gcc to avoid symbol name collision when a const is generated multiple times doesn't work if multiple CGUs decide to codegen the same symbol for a promoted constant.

The proper fix I think is to let backend mint new symbol names, rather than construct a fake `Instance` and use its symbol name like the current approach. This does mean that I need to create a new mechanism for backend to mint new global symbol names.

nbdd0121 committed:
Give codegen units symbol names that backend can use

nbdd0121 commented:
I've reworked how symbol name is generated and stop using the mangled name of the const, due to the outlined symbol name conflict issue above. This also resolves the concern https://github.com/rust-lang/rust/pull/138618#discussion_r3061331232.

The new approach generates an unique symbol name prefix for each CGU, and then the codegen backends can use this symbol prefix to create helper symbols that they need internally. I think this can be useful longer term as well. The generated prefix is demangled to something like `crate_name::{shim:cgu#1}` for non-incremental build and  `crate_name::{shim:cgu#0}::<"incremental CGU name">` for incremental build.

EDIT: Due to two pushes the range-diff above is not the full picture, [this](https://triagebot.infra.rust-lang.org/gh-range-diff/rust-lang/rust/cf1817bc6ecd2d14ca492247c804bad31948dd56..7c877c9f0a8386b02ff1b71d2a3409533ad6e070/9d862dd2c17b2dc90e9f623bb9279a8c4090b311..ff3f3c03cfcf3800e133754094a19e682f0ae3b0) is the full range diff.

Darksonn commented:
Is the cgu index still included in the symbol name when the number of cgus is 1?

nbdd0121 commented:
It'll still demangle to things like `crate_name::{shim:cgu#0}` even with one CGU. However, v0 mangling will omit disambiguator completely if it is of value 0, so the resulting symbol name is just going to be something like `_RNSCs<stable crate ID>_10crate_name3cgu`

nbdd0121 commented:
@rustbot ready

bjorn3 commented on compiler/rustc_hir_typeck/src/inline_asm.rs:
```diff
@@ -548,15 +548,46 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
+                            }
+                        }
+                        ty::RawPtr(pointee, _) | ty::Ref(_, pointee, _)
+                            if self.is_thin_ptr_ty(*pointee) =>
```
Should we allow `extern { type Foo; }` like this method does?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cde3f8aee5a30928872e2438e3f18238a3fa306c..61a45c60050d6ef3eb619fd85795cbbb3cbf4701)*

nbdd0121 replied:
I think it's a valid use case to pass extern pointers to assembly? Cc @Darksonn

bjorn3 replied:
r=me with this question decided one way or the other.

Darksonn replied:
The main concern I see is `&CStr`. As the RFC points out, the main benefit of disallowing wide pointers is avoiding confusion about whether `const &str` means interpolation or pointer.

Darksonn replied:
On the other hand, for basically every other `extern { type Foo; }` I think we do want it to work.

Darksonn replied:
Thinking more about it, I don't think I'm that worried about `&CStr` being confused like this. So I think allowing extern types is fine.

bjorn3 commented on compiler/rustc_symbol_mangling/src/v0.rs:
```diff
@@ -90,6 +90,42 @@ pub(super) fn mangle<'tcx>(
     std::mem::take(&mut p.out)
 }
 
+pub fn mangle_cgu<'tcx>(tcx: TyCtxt<'tcx>, krate: CrateNum, cgu_name: Result<u64, &str>) -> String {
```
Maybe `rustc_data_stuctures::either::Either` instead of `Result`?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cde3f8aee5a30928872e2438e3f18238a3fa306c..61a45c60050d6ef3eb619fd85795cbbb3cbf4701)*

bjorn3 commented on compiler/rustc_hir_typeck/src/expr.rs:
Someone else will have to review the typeck changes.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cde3f8aee5a30928872e2438e3f18238a3fa306c..61a45c60050d6ef3eb619fd85795cbbb3cbf4701)*

nbdd0121 replied:
FWIW this part isn't changed since @saethlin's last review in https://github.com/rust-lang/rust/pull/138618#issuecomment-4218731935.

bjorn3 commented on compiler/rustc_codegen_ssa/src/common.rs:
```diff
@@ -170,7 +167,34 @@ pub fn asm_const_to_str<'tcx>(
         },
-        _ => span_bug!(sp, "asm const has bad type {}", ty_and_layout.ty),
+        // For unsigned integers or pointers without provenance, just print the unsigned value
+        _ => value.to_string(),
```
Unsigned ints are already handled above. For pointers without provenance we can have an explicit arm. I would prefer keeping this bug!() so we don't forget to update this if we ever allow more types like floats.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/cde3f8aee5a30928872e2438e3f18238a3fa306c..61a45c60050d6ef3eb619fd85795cbbb3cbf4701)*

Darksonn commented on tests/ui/asm/invalid-const-operand.rs:
```diff
@@ -13,21 +15,25 @@ global_asm!("{}", const 0i128);
-        //~^ ERROR invalid type for `const` operand
         asm!("{}", const &0);
+        asm!("{}", const b"Foo".as_slice());
         //~^ ERROR invalid type for `const` operand
```
I'd also like to see `&str` and `&CStr` types included in the test suite.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/138618/c188565812a9b414ddf02f3c7d89e84f1ba53f7d..0e139d3adf1500f366ddda40cd5c5a98a2c799a7)*

nbdd0121 replied:
Added

nbdd0121 committed:
Expand invalid-const-operand test case to cover &str and &CStr

Use edition 2021 so c-string literals are supported.
